fawltydeps/extract_declared_dependencies.py

```diff
@@ -340,52 +340,84 @@ def extract_declared_dependencies_from_path(
+    assert src.path.is_file()  # sanity check
+    if parser_choice is not None:
+        parser = PARSER_CHOICES[parser_choice]
+        if not parser.applies_to_path(src.path):
+            logger.warning(
+                f"Manually applying parser '{parser_choice}' to dependencies: {src.path}"
+            )
     else:
-        raise UnparseablePathException(
-            ctx="Dependencies declaration path is neither dir nor file", path=path
-        )
+        choice_and_parser = first_applicable_parser(src.path)
+        if choice_and_parser is None:  # nothing found. SHOULD NOT HAPPEN!
+            raise UnparseablePathException(
+                ctx="Parsing given dependencies path isn't supported", path=src.path
+            )
+        parser = choice_and_parser[1]
+    yield from parser.execute(src.path)
 
 
-def extract_declared_dependencies(
-    paths: Set[Path], parser_choice: Optional[ParserChoice] = None
+def parse_sources(
+    sources: Iterable[DepsSource], parser_choice: Optional[ParserChoice] = None
 ) -> Iterator[DeclaredDependency]:
     """Extract dependencies (package names) from supported file types.
 
-    Pass a list of paths from which to discover and parse dependency declarations.
+    Pass sources from which to parse dependency declarations.
     """
+    for source in sources:
+        yield from parse_source(source, parser_choice=parser_choice)
 
-    for path in paths:
-        yield from extract_declared_dependencies_from_path(
-            path, parser_choice=parser_choice
-        )
+
+def is_valid_deps_source(
+    path: Path,
+    parser_choice: Optional[ParserChoice] = None,
+    filter_by_parser: bool = False,
+) -> Optional[DepsSource]:
+    """Check if the given file path is a valid source for parsing declared deps.
+
+    - Return the given path as a DepsSource object iff it is a file that we know
+      how to parse.
+    - Return None if this is a directory that must be traversed further to find
+      parseable files within.
+    - Raise UnparseablePathException if the given path cannot be parsed.
+
+    The given 'parser_choice' and 'filter_by_parser' change which files we
+    consider valid sources:
+
+    - If parser_choice is None, we will consider any file path that matches one
+      of the PARSER_CHOICES above a valid file.
+    - If parser_choice is not None, and filter_by_parser is False, we assume
+      that the given file path was explicitly passed by the user, and meant to
+      be parsed with the given parser_choice, _even_ when the file path does not
+      nominally match this parser choice: We return the file path as a valid
+      source regardless of its file name/suffix.
+    - If parser_choice is not None, and filter_by_parser is True, we assume that
+      the given file path was encountered while walking a directory passed by
+      the user. In this case, we will only consider the file valid if it matches
+      the given parser choice.
```

Comment:
I've shortened this description in commit 8a36a369c8a3a175f435b4ce40403cbbdf103e7d, but tried to keep the main points:
```
    The given 'parser_choice' and 'filter_by_parser' change which files we
    consider valid sources: With parser_choice=None, any file path that matches
    one of the PARSER_CHOICES above is considered a valid file. Otherwise,
    filter_by_parser is used to either consider only files that match the given
    parser_choice (filter_by_parser=True), or to force all files to be parsed as
    if they matched the given parser_choice (filter_by_parser=False).
```